helix-term/src/commands/lsp.rs

```diff
@@ -424,81 +497,117 @@ impl ui::menu::Item for lsp::CodeActionOrCommand {
-                })
-                .map(|diag| diagnostic_to_lsp_diagnostic(doc.text(), diag, offset_encoding))
-                .collect(),
-            only: None,
-        },
-    );
+    let mut requests = Vec::new();
 
-    cx.callback(
-        future,
-        move |editor, compositor, response: Option<lsp::CodeActionResponse>| {
-            let actions = match response {
-                Some(a) => a,
-                None => return,
-            };
-            if actions.is_empty() {
-                editor.set_status("No code actions available");
-                return;
-            }
+    for language_server in doc.language_servers_with_feature(LanguageServerFeature::CodeAction) {
+        let offset_encoding = language_server.offset_encoding();
+        let range = range_to_lsp_range(doc.text(), selection_range, offset_encoding);
+
+        requests.push((
+            language_server.code_actions(
+                doc.identifier(),
+                range,
+                // Filter and convert overlapping diagnostics
+                lsp::CodeActionContext {
+                    diagnostics: doc
+                        .diagnostics()
+                        .iter()
+                        .filter(|&diag| {
+                            selection_range
+                                .overlaps(&helix_core::Range::new(diag.range.start, diag.range.end))
+                        })
+                        .map(|diag| diagnostic_to_lsp_diagnostic(doc.text(), diag, offset_encoding))
+                        .collect(),
+                    only: None,
+                },
+            ),
+            offset_encoding,
+            language_server.id(),
+        ));
+    }
+
+    for (future, offset_encoding, lsp_id) in requests {
+        let code_actions_menu_open = code_actions_menu_open.clone();
+
+        cx.callback(
+            future,
+            move |editor, compositor, response: Option<lsp::CodeActionResponse>| {
+                let actions = match response {
+                    Some(a) => a
+                        .into_iter()
+                        .map(|a| (a, offset_encoding))
+                        .collect::<Vec<_>>(),
+                    None => return,
+                };
+
+                let mut code_actions_menu_open = code_actions_menu_open.lock().unwrap();
 
-            let mut picker = ui::Menu::new(actions, (), move |editor, code_action, event| {
-                if event != PromptEvent::Validate {
+                if actions.is_empty() && !*code_actions_menu_open {
+                    editor.set_status("No code actions available");
                     return;
                 }
 
-                // always present here
-                let code_action = code_action.unwrap();
+                let code_actions_menu = compositor.find_id::<Popup<
+                    ui::Menu<(lsp::CodeActionOrCommand, OffsetEncoding)>,
+                >>("code-action");
```

Comment:
🤯😭 could we maybe make a type alias for this?

Reply:
Usually I agree with this, but since it's only used there, and the name of the variable hints what is found, I think it doesn't really help with readability IMHO. I can still create a type alias if you or others think it really helps with readability.

Reply:
Ok just tried it out, it saves one line (yay^^) and looks slightly neater